src/Control/Monad/Freer/Resource.hs

```diff
@@ -0,0 +1,262 @@
+-- resources from exiting their region scope.
+data L (n :: Nat) k
+
+
+------------------------------------------------------------------------------
+-- | Helper function to build region constructs for resources of type 'res'.
+handleRegionRelay :: forall res effs a
+                   . ( SafeForRegion res effs
+                     , Eq (HiddenResource res)
+                     )
+                  => -- | Strategy to acquire a 'res'.
+                     (ResourceCtor res -> Eff effs (HiddenResource res))
+                     -- | Strategy to release a 'res'.
+                  -> (HiddenResource res -> Eff effs ())
+                     -- | Strategy for catching other effects.
+                  -> (forall (b :: *)
+                         . -- | Action to release all resources.
+                           Eff effs ()
+                           -- | Action to ignore this effect.
+                        -> (Union effs b -> Eff effs a)
+                           -- | The effect being caught.
+                        -> Union effs b
+                        -> Eff effs a
+                     )
+                     -- | A region in which we can allocate 'res's.
+                  -> (forall s. Eff (RegionEff res (L (Length effs) s) ': effs) a)
+                  -> Eff effs a
+handleRegionRelay acquireM releaseM catchM = loop []
+  where
+    releaseAll = mapM_ releaseM
+
+    loop :: [HiddenResource res]
+         -> Eff (RegionEff res (L (Length effs) s) ': effs) a
+         -> Eff effs a
+    loop rs (Val x) = releaseAll rs >> return x
+    loop rs (E u q) =
+      case decomp u of
+        Right (RENew ctor) -> do
+          r <- acquireM ctor
+          k (r : rs) $ Resource r
+
+        Right (REForget (Resource r)) ->
+          k (delete r rs) ()
+
+        Right (REAcquire (Resource r)) ->
+          k (bool (r : rs) rs $ elem r rs) $ Resource r
+
+        Left u' ->
+          catchM (releaseAll rs) ignore u'
+
+      where
+        ignore u' = E u' . tsingleton $ k rs
+        k s = qComp q (loop s)
+
+
+------------------------------------------------------------------------------
+-- | Combinator to be used with 'handleRegionRelay' describing that no extra
+-- effects need being caught.
+catchNothing :: Eff effs ()
+             -> (Union effs b -> Eff effs a)
+             -> Union effs b
+             -> Eff effs a
+catchNothing = const id
+
+
+------------------------------------------------------------------------------
+-- | Example region for acquiring file 'Handle's.
+fileHandleRegion :: forall r a
+                  . ( Member IO r
+                    , SafeForRegion Handle r
+                    , Member (Exc SomeException) r
+                    )
+                 => (forall s. Eff (RegionEff Handle (L (Length r) s) ': r) a)
+                 -> Eff r a
+fileHandleRegion = handleRegionRelay (send . fmap HandleResource . uncurry openFile)
+                                     (send . close . unHandleResource)
```

Comment:
Haven't yet had a chance to walk through this whole file... what happens if an async exception is thrown to this thread while within the region?